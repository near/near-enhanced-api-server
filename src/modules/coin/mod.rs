use actix_web::{HttpResponse, error};
use actix_web_validator::{Error, PathConfig};
use paperclip::actix::web;

mod data_provider;
mod resources;
mod schemas;

#[derive(serde::Serialize)]
pub struct ValidationErrorJsonPayload {
    pub message: String,
    pub fields: Vec<String>,
}

/// Custom error handler
impl From<&validator::ValidationErrors> for ValidationErrorJsonPayload {
    fn from(error: &validator::ValidationErrors) -> Self {
        ValidationErrorJsonPayload {
            message: "Validation error".to_owned(),
            fields: error.field_errors().iter().map(|(field, _)| field.to_string()).collect(),
        }
    }
}

pub(crate) fn register_services(app: &mut web::ServiceConfig) {
    app
        .app_data(PathConfig::default().error_handler(|err, _| {
            let json_error = match &err {
                Error::Validate(error) => ValidationErrorJsonPayload::from(error),
                _ => ValidationErrorJsonPayload { message: err.to_string(), fields: Vec::new() },
            };
            error::InternalError::from_response(err, HttpResponse::Conflict().json(json_error)).into()
        }))
        .service(
        web::resource("/accounts/{account_id}/coins/NEAR")
            .route(web::get().to(resources::get_near_balance)),
    )
    .service(
        web::resource("/accounts/{account_id}/coins")
            .route(web::get().to(resources::get_coin_balances)),
    )
    .service(
        web::resource("/accounts/{account_id}/coins/{contract_account_id}")
            .route(web::get().to(resources::get_coin_balances_by_contract)),
    )
    .service(
        web::resource("/accounts/{account_id}/coins/NEAR/history")
            .route(web::get().to(resources::get_near_history)),
    )
    .service(
        web::resource("/accounts/{account_id}/coins/{contract_account_id}/history")
            .route(web::get().to(resources::get_coin_history)),
    )
    .service(
        web::resource("/nep141/metadata/{contract_account_id}")
            .route(web::get().to(resources::get_ft_contract_metadata)),
    );
}
