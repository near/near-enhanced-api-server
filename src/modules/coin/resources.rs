use paperclip::actix::{
    api_v2_operation,
    web::{self, Json},
};

use super::{data_provider, schemas};
use crate::{db_helpers, errors, modules, types};

#[api_v2_operation(tags(Coins))]
/// Get user's NEAR balance
///
/// This endpoint returns the NEAR balance of the given `account_id`
/// at the given `block_timestamp_nanos`/`block_height`.
pub async fn get_near_balance(
    pool: web::Data<sqlx::Pool<sqlx::Postgres>>,
    rpc_client: web::Data<near_jsonrpc_client::JsonRpcClient>,
    _: crate::types::pagoda_api_key::PagodaApiKey,
    request: actix_web_validator::Path<schemas::BalanceRequest>,
    block_params: web::Query<types::query_params::BlockParams>,
) -> crate::Result<Json<schemas::NearBalanceResponse>> {
    let block = db_helpers::checked_get_block(&pool, &block_params).await?;
    modules::check_account_exists(&rpc_client, &request.account_id.0, block.height).await?;

    Ok(Json(
        data_provider::get_near_balance(&pool, &block, &request.account_id.0).await?,
    ))
}

#[api_v2_operation(tags(Coins))]
/// Get user's coin balances
///
/// This endpoint returns all the countable coin balances (including NEAR, FTs)
/// of the given `account_id`, at the given `block_timestamp_nanos`/`block_height`.
/// todo it's not great, we can't keep sorting + pagination when we add MT here
/// NEAR balance always goes first, then the balances are sorted alphabetically by `contract_account_id`.
///
/// You can also provide `limit` (max 100) and `after_contract_account_id`.
/// `after_contract_account_id` is the way to paginate through the values.
///
/// Some pagination notes: leave `after_contract_account_id` empty at your first request;
/// then, take `contract_account_id` of the last provided item in your previous response.
///
/// **Limitations**
/// * We are in the process of supporting Multi Token balances.
pub async fn get_coin_balances(
    pool: web::Data<sqlx::Pool<sqlx::Postgres>>,
    rpc_client: web::Data<near_jsonrpc_client::JsonRpcClient>,
    _: crate::types::pagoda_api_key::PagodaApiKey,
    request: actix_web_validator::Path<schemas::BalanceRequest>,
    block_params: web::Query<types::query_params::BlockParams>,
    limit_params: web::Query<types::query_params::LimitParams>,
) -> crate::Result<Json<schemas::CoinBalancesResponse>> {
    let mut limit = types::query_params::checked_get_limit(limit_params.limit)?;
    let block = db_helpers::checked_get_block(&pool, &block_params).await?;
    modules::check_account_exists(&rpc_client, &request.account_id.0, block.height).await?;

    let mut balances: Vec<schemas::Coin> = vec![];
    balances.push(
        data_provider::get_near_balance(&pool, &block, &request.account_id.0)
            .await?
            .into(),
    );
    limit -= 1;

    if limit > 0 {
        let ft_balances = &mut data_provider::get_coin_balances(
            &pool,
            &rpc_client,
            &block,
            &request.account_id.0,
            limit,
        )
        .await?;
        balances.append(ft_balances);
    }

    Ok(Json(schemas::CoinBalancesResponse {
        balances,
        block_timestamp_nanos: types::U64::from(block.timestamp),
        block_height: types::U64::from(block.height),
    }))
}

#[api_v2_operation(tags(Coins))]
/// Get user's coin balances by contract
///
/// This endpoint returns all the countable coin balances of the given `account_id`,
/// for the given `contract_account_id` and `block_timestamp_nanos`/`block_height`.
/// For FT contracts, the response has only 1 item in the list.
/// For MT contracts, there could be several balances (MT support is still under development).
/// todo change interface for ft/mt
///
/// **Limitations**
/// * We are in the process of supporting Multi Token balances.
pub async fn get_coin_balances_by_contract(
    pool: web::Data<sqlx::Pool<sqlx::Postgres>>,
    rpc_client: web::Data<near_jsonrpc_client::JsonRpcClient>,
    _: crate::types::pagoda_api_key::PagodaApiKey,
    request: actix_web_validator::Path<schemas::BalanceByContractRequest>,
    block_params: web::Query<types::query_params::BlockParams>,
) -> crate::Result<Json<schemas::CoinBalancesResponse>> {
    if request.contract_account_id.to_string() == "near" {
        return Err(errors::ErrorKind::InvalidInput(
            "For native balance, please use NEAR (uppercase)".to_string(),
        )
        .into());
    }
    let block = db_helpers::checked_get_block(&pool, &block_params).await?;
    modules::check_account_exists(&rpc_client, &request.account_id.0, block.height).await?;

    let balances = data_provider::get_coin_balances_by_contract(
        &rpc_client,
        &block,
        &request.contract_account_id.0,
        &request.account_id.0,
    )
    .await?;

    Ok(Json(schemas::CoinBalancesResponse {
        balances,
        block_timestamp_nanos: types::U64::from(block.timestamp),
        block_height: types::U64::from(block.height),
    }))
}

#[api_v2_operation(tags(Coins))]
/// Get user's NEAR history
///
/// This endpoint returns the history of native NEAR operations,
/// for the given `account_id`, going from the latest to the earliest events.
/// For the next page, use `after_event_index` and pass there the index from the last event from the previous page.
///
/// **Limitations**
/// We are in the process of collecting the data. We support only last 3 months history for now.
pub async fn get_near_history(
    pool: web::Data<sqlx::Pool<sqlx::Postgres>>,
    pool_balances: web::Data<db_helpers::DBWrapper>,
    _: crate::types::pagoda_api_key::PagodaApiKey,
    request: actix_web_validator::Path<schemas::BalanceRequest>,
    pagination_params: web::Query<types::query_params::PaginationParams>,
) -> crate::Result<Json<schemas::HistoryResponse>> {
    let pagination = modules::checked_get_pagination_params(&pagination_params).await?;
    let block = db_helpers::checked_get_block_from_pagination(&pool, &pagination).await?;
    // we don't need to check whether account exists. If not, we can just return the empty history

    Ok(Json(schemas::HistoryResponse {
        history: data_provider::get_near_history(
            &pool_balances.pool,
            &request.account_id.0,
            &block,
            &pagination,
        )
        .await?,
        block_timestamp_nanos: types::U64::from(block.timestamp),
        block_height: types::U64::from(block.height),
    }))
}

#[api_v2_operation(tags(Coins))]
/// Get user's coin history by contract
///
/// This endpoint returns the history of coin operations (FT, other standards)
/// for the given `account_id`, `contract_id`, going from the latest to the earliest events.
/// For the next page, use `after_event_index` and pass there the index from the last event from the previous page.
///
/// **Limitations**
/// * If the history does not match with the balances from RPC, we do not support the history (see Trusted/Inconsistent Contracts section below).
/// * We are in the process of supporting Multi Token history.
///
/// **Trusted/Inconsistent Contracts**
/// The contracts produce [the balance changing events](https://nomicon.io/Standards/Tokens/FungibleToken/Event).
/// The contract is marked as inconsistent if its events do not match with the balance taken from RPC.
/// The contract is marked as trusted otherwise.
/// More info could be found [here](https://github.com/near/near-indexer-events#my-contract-produces-eventstheres-a-custom-legacy-logic-for-my-contract-but-the-enhanced-api-still-ignores-me-why).
pub async fn get_coin_history(
    pool: web::Data<sqlx::Pool<sqlx::Postgres>>,
    pool_balances: web::Data<db_helpers::DBWrapper>,
    rpc_client: web::Data<near_jsonrpc_client::JsonRpcClient>,
    _: crate::types::pagoda_api_key::PagodaApiKey,
    request: actix_web_validator::Path<schemas::HistoryRequest>,
    pagination_params: web::Query<types::query_params::PaginationParams>,
) -> crate::Result<Json<schemas::HistoryResponse>> {
    if request.contract_account_id.to_string() == "near" {
        return Err(errors::ErrorKind::InvalidInput(
            "For native history, please use NEAR (uppercase)".to_string(),
        )
        .into());
    }

    let pagination = modules::checked_get_pagination_params(&pagination_params).await?;
    let block = db_helpers::checked_get_block_from_pagination(&pool, &pagination).await?;
    // we don't need to check whether account exists. If not, we can just return the empty history

    Ok(Json(schemas::HistoryResponse {
        history: data_provider::get_coin_history(
            &pool,
            &pool_balances.pool,
            &rpc_client,
            &request.contract_account_id.0,
            &request.account_id.0,
            &block,
            &pagination,
        )
        .await?,
        block_timestamp_nanos: types::U64::from(block.timestamp),
        block_height: types::U64::from(block.height),
    }))
}

#[api_v2_operation(tags(Coins))]
/// Get FT contract metadata
///
/// This endpoint returns the metadata for a given FT contract and `timestamp`/`block_height`.
///
/// **Limitations**
/// * For now, we support only FT contracts that implement the Events NEP standard.
///   We are working on a solution to support other FT contracts, including `wrap.near` and bridged tokens.
pub async fn get_ft_contract_metadata(
    pool: web::Data<sqlx::Pool<sqlx::Postgres>>,
    rpc_client: web::Data<near_jsonrpc_client::JsonRpcClient>,
    _: crate::types::pagoda_api_key::PagodaApiKey,
    request: actix_web_validator::Path<schemas::ContractMetadataRequest>,
    block_params: web::Query<types::query_params::BlockParams>,
) -> crate::Result<Json<schemas::FtContractMetadataResponse>> {
    let block = db_helpers::checked_get_block(&pool, &block_params).await?;

    Ok(Json(schemas::FtContractMetadataResponse {
        metadata: data_provider::get_ft_contract_metadata(
            &rpc_client,
            request.contract_account_id.0.clone(),
            block.height,
        )
        .await?,
        block_timestamp_nanos: types::U64::from(block.timestamp),
        block_height: types::U64::from(block.height),
    }))
}
