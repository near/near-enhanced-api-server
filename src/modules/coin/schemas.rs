use paperclip::actix::Apiv2Schema;
use validator::{Validate, ValidationError};

use crate::types;

// *** Requests ***

// move to coins
#[derive(Validate, Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct BalanceRequest {
    #[validate(custom = "near_primitives::types::AccountId::validate")]
    pub account_id: types::AccountId,
}

#[derive(Validate, Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct BalanceByContractRequest {
    #[validate(custom = "near_primitives::types::AccountId::validate")]
    pub account_id: types::AccountId,
    #[validate(custom = "near_primitives::types::AccountId::validate")]
    pub contract_account_id: types::AccountId,
}

#[derive(Validate, Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct HistoryRequest {
    #[validate(custom = "near_primitives::types::AccountId::validate")]
    pub account_id: types::AccountId,
    pub contract_account_id: types::AccountId,
}

// duplicate in each folder
#[derive(Validate, Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct ContractMetadataRequest {
    #[validate(custom = "near_primitives::types::AccountId::validate")]
    pub contract_account_id: types::AccountId,
}

// *** Responses ***

#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct NearBalanceResponse {
    /// Sum of staked and nonstaked balances
    pub balance: types::U128,
    pub metadata: CoinMetadata,
    pub block_timestamp_nanos: types::U64,
    pub block_height: types::U64,
}

/// This response gives the information about all the available balances for the user.
/// The answer gives the list of NEAR, FT balances, could be used for Multi Tokens.
/// For MTs and other standards, balances could have multiple entries for one contract.
#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct CoinBalancesResponse {
    pub balances: Vec<Coin>,
    pub block_timestamp_nanos: types::U64,
    pub block_height: types::U64,
}

/// This response provides the coin history (NEAR or by contract).
#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct HistoryResponse {
    pub history: Vec<HistoryItem>,
    pub block_timestamp_nanos: types::U64,
    pub block_height: types::U64,
}

#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct FtContractMetadataResponse {
    pub metadata: FtContractMetadata,
    pub block_timestamp_nanos: types::U64,
    pub block_height: types::U64,
}

// ---

/// This type describes general coin information.
/// It is used for NEAR, FT balances, could be used for Multi Tokens.
///
/// For MTs and other standards, we could have multiple coins for one contract.
/// For NEAR and FTs, coin_metadata contains general metadata (the only available option, though).
#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct Coin {
    /// "nearprotocol" for NEAR, "nep141" for FT
    pub standard: String,
    pub balance: types::U128,
    /// null for NEAR, not null otherwise
    pub contract_account_id: Option<types::AccountId>,
    pub metadata: CoinMetadata,
    // TODO PHASE 1 (idea) I think it would be great to add here the info about last update moment. Timestamp, later also index
    // I'm already doing it at NftCount
}

/// This type describes the history of coin movements for the given user.
/// Coins could be NEAR, FT, it could be also later used for Multi Tokens.
#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct HistoryItem {
    // TODO PHASE 2 add index here
    // pub index: types::U128,
    // TODO PHASE 1 (idea) do we want to add here tx_hash/receipt_id? We may want to add it at many places
    pub involved_account_id: Option<types::AccountId>,
    pub delta_balance: types::I128,
    pub balance: types::U128,
    pub cause: String,
    pub status: String,
    pub coin_metadata: CoinMetadata,
    pub block_timestamp_nanos: types::U64,
    // TODO PHASE 2 add this when we have all the data in the same DB. Now we can't join with blocks
    // pub block_height: types::U64,
}

/// This type describes general Metadata info, collecting the most important fields from different standards in the one format.
/// `decimals` may contain `0` if it's not applicable (e.g. if it's general MT metadata)
#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct CoinMetadata {
    pub name: String,
    pub symbol: String,
    pub icon: Option<String>,
    pub decimals: u8,
}

/// The type for FT Contract Metadata. Inspired by
/// https://nomicon.io/Standards/Tokens/FungibleToken/Metadata
#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct FtContractMetadata {
    pub spec: String,
    pub name: String,
    pub symbol: String,
    pub icon: Option<String>,
    pub reference: Option<String>,
    pub reference_hash: Option<String>,
    pub decimals: u8,
}

pub fn validate(account_id: &str) -> Result<(), ValidationError> {
    Err(ValidationError::new("something"))
}
