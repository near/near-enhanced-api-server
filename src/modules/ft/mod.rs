use paperclip::actix::web;

mod data_provider;
mod resources;
mod schemas;

pub(crate) fn register_services(app: &mut web::ServiceConfig) {
    // I'll drop these comments, leaving it only because it's easier for the review
    // app.service(
    //     web::resource("/accounts/{account_id}/NEAR")
    //         .route(web::get().to(resources::get_near_balance)),
    // )
    app.service(
        web::resource("/accounts/{account_id}/FT").route(web::get().to(resources::get_ft_balances)),
    )
    .service(
        web::resource("/accounts/{account_id}/FT/{contract_account_id}")
            .route(web::get().to(resources::get_ft_balance_by_contract)),
    )
    // .service(
    //     web::resource("/accounts/{account_id}/NEAR/history")
    //         .route(web::get().to(resources::get_near_history)),
    // )
    .service(
        web::resource("/accounts/{account_id}/FT/{contract_account_id}/history")
            .route(web::get().to(resources::get_ft_history)),
    )
    .service(
        web::resource("/nep141/metadata/{contract_account_id}")
            .route(web::get().to(resources::get_ft_metadata)),
    );
}
