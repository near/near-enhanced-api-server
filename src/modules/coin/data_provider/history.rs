use num_traits::Signed;
use std::str::FromStr;

use crate::modules::coin;
use crate::{db_helpers, errors, types, BigDecimal};

pub(crate) async fn get_near_history(
    balances_pool: &sqlx::Pool<sqlx::Postgres>,
    account_id: &near_primitives::types::AccountId,
    block: &db_helpers::Block,
    pagination: &types::query_params::Pagination,
) -> crate::Result<Vec<coin::schemas::HistoryItem>> {
    // todo
    let query = if pagination.after_event_index.is_some() {
        r"
        SELECT
            event_index,
            involved_account_id,
            delta_nonstaked_amount + delta_staked_amount delta_balance,
            absolute_nonstaked_amount + absolute_staked_amount balance,
            cause,
            status,
            block_timestamp block_timestamp_nanos,
            block_height
        FROM balance_changes
        WHERE affected_account_id = $1 AND event_index < $2::numeric(38, 0)
        ORDER BY event_index DESC
        LIMIT $4::numeric(20, 0)
    "
    } else {
        r"
        SELECT
            event_index,
            involved_account_id,
            delta_nonstaked_amount + delta_staked_amount delta_balance,
            absolute_nonstaked_amount + absolute_staked_amount balance,
            cause,
            status,
            block_timestamp block_timestamp_nanos,
            block_height
        FROM balance_changes
        WHERE affected_account_id = $1 AND block_timestamp <= $3::numeric(20, 0)
        ORDER BY event_index DESC
        LIMIT $4::numeric(20, 0)
    "
    };

    let history_info = db_helpers::select_retry_or_panic::<super::models::CoinHistoryInfo>(
        balances_pool,
        query,
        &[
            account_id.to_string(),
            pagination.after_event_index.unwrap_or(0).to_string(),
            block.timestamp.to_string(),
            pagination.limit.to_string(),
        ],
    )
    .await?;

    let mut result: Vec<coin::schemas::HistoryItem> = vec![];
    for history in history_info {
        let involved_account_id: Option<types::AccountId> =
            if let Some(account_id) = history.involved_account_id {
                Some(near_primitives::types::AccountId::from_str(&account_id)?.into())
            } else {
                None
            };

        result.push(coin::schemas::HistoryItem {
            event_index: types::numeric::to_u128(&history.event_index)?.into(),
            involved_account_id,
            delta_balance: history.delta_balance.to_string(),
            balance: types::numeric::to_u128(&history.balance)?.into(),
            cause: history.cause,
            status: history.status,
            coin_metadata: super::get_near_metadata(),
            block_timestamp_nanos: types::numeric::to_u64(&history.block_timestamp_nanos)?.into(),
            block_height: types::numeric::to_u64(&history.block_height)?.into(),
        });
    }
    Ok(result)
}

pub(crate) async fn get_coin_history(
    pool: &sqlx::Pool<sqlx::Postgres>,
    pool_events: &sqlx::Pool<sqlx::Postgres>,
    rpc_client: &near_jsonrpc_client::JsonRpcClient,
    contract_id: &near_primitives::types::AccountId,
    account_id: &near_primitives::types::AccountId,
    block: &db_helpers::Block,
    pagination: &types::query_params::Pagination,
) -> crate::Result<Vec<coin::schemas::HistoryItem>> {
    let balance_u128 = super::balance::get_ft_balance_by_contract(
        rpc_client,
        contract_id.clone(),
        account_id.clone(),
        block.height,
    )
    .await?;
    let mut current_balance = BigDecimal::from_str(&balance_u128.to_string()).map_err(|e| {
        errors::ErrorKind::InternalError(format!("Failed to parse BigDecimal from u128: {}", e))
    })?;

    let metadata = coin::schemas::CoinMetadata::from(
        super::metadata::get_ft_contract_metadata(rpc_client, contract_id.clone(), block.height)
            .await?,
    );

    // We don't have absolute_value in the DB.
    // We ask RPC for the first absolute_value, fill the in-between values with deltas, check by asking RPC for the last absolute_value and comparing the results.
    // If it matches, everything is good. If not, we return the error.

    // Problem: we can ask RPC only for the balance at the end of the block, while we may have the lines starting and ending in the middle of the block.
    // That's why we need to take slightly more lines so that we can match it with the RPC, and then filter these lines.
    // Potential issue here: the DB may contain only the part of the most fresh block (we may be in the middle of the writing process)
    // It can be easily solved by ignoring the most fresh block, but it will increase the lag between the response and the current blockchain state.
    // Since we anyway use transactional DB which guarantees that write process goes atomically, I don't want to do anything with that.
    // But, if we meet such issues in production, we may consider cutting the latest block.
    let query = if pagination.after_event_index.is_some() {
        r"
        WITH original_query as (
            SELECT
                event_index,
                involved_account_id,
                delta_amount,
                cause,
                status,
                block_timestamp,
                block_height
            FROM coin_events
            WHERE contract_account_id = $1
                AND affected_account_id = $2
                AND event_index < $3::numeric(38, 0)
            ORDER BY event_index desc
            LIMIT $5::numeric(20, 0)
        ), timestamps as (
            SELECT
                min(block_timestamp) min_block_timestamp,
                max(block_timestamp) max_block_timestamp
            FROM original_query
        )
        SELECT
            event_index,
            involved_account_id,
            delta_amount delta_balance,
            0::numeric(40, 0) balance, -- fill it later
            cause,
            status,
            block_timestamp block_timestamp_nanos,
            block_height
        FROM coin_events, timestamps
        WHERE contract_account_id = $1
            AND affected_account_id = $2
            AND block_timestamp >= min_block_timestamp
            AND block_timestamp <= max_block_timestamp
        ORDER BY event_index desc
    "
    } else {
        r"
        WITH original_query as (
            SELECT
                event_index,
                involved_account_id,
                delta_amount,
                cause,
                status,
                block_timestamp,
                block_height
            FROM coin_events
            WHERE contract_account_id = $1
                AND affected_account_id = $2
                AND block_timestamp <= $4::numeric(20, 0)
            ORDER BY event_index desc
            LIMIT $5::numeric(20, 0)
        ), timestamps as (
            SELECT
                min(block_timestamp) min_block_timestamp,
                max(block_timestamp) max_block_timestamp
            FROM original_query
        )
        SELECT
            event_index,
            involved_account_id,
            delta_amount delta_balance,
            0::numeric(40, 0) balance, -- fill it later
            cause,
            status,
            block_timestamp block_timestamp_nanos,
            block_height
        FROM coin_events, timestamps
        WHERE contract_account_id = $1
            AND affected_account_id = $2
            AND block_timestamp >= min_block_timestamp
            AND block_timestamp <= max_block_timestamp
        ORDER BY event_index desc
    "
    };
    let history =
        db_helpers::select_retry_or_panic::<coin::data_provider::models::CoinHistoryInfo>(
            pool_events,
            query,
            &[
                contract_id.to_string(),
                account_id.to_string(),
                pagination.after_event_index.unwrap_or(0).to_string(),
                block.timestamp.to_string(),
                pagination.limit.to_string(),
            ],
        )
        .await?;

    let mut result: Vec<coin::schemas::HistoryItem> = vec![];
    for db_info in history {
        let balance = current_balance.clone();

        if db_info.status == "SUCCESS" {
            current_balance -= db_info.delta_balance.clone();
            if current_balance.is_negative() {
                return Err(errors::ErrorKind::InternalError(format!(
                    "History is not supported for account {}. Contract {} provides inconsistent events which lead to negative balance",
                    account_id, contract_id
                ))
                    .into());
            }
        }

        let involved_account_id = match db_info.involved_account_id {
            Some(id) => Some(types::AccountId::from_str(&id)?),
            None => None,
        };
        let event_index = types::numeric::to_u128(&db_info.event_index)?;

        // We collect slightly more lines that we were asked for, because we can make RPC calls only at the end of the block
        // First clause filters latest redundant lines, second clause filters earliest redundant lines
        if pagination
            .after_event_index
            .map_or_else(|| true, |index| index > event_index)
            && result.len() < pagination.limit as usize
        {
            result.push(coin::schemas::HistoryItem {
                event_index: event_index.into(),
                cause: db_info.cause.clone(),
                involved_account_id,
                delta_balance: db_info.delta_balance.to_string(),
                balance: types::numeric::to_u128(&balance)?.into(),
                coin_metadata: metadata.clone(),
                block_timestamp_nanos: types::numeric::to_u64(&db_info.block_timestamp_nanos)?
                    .into(),
                block_height: types::numeric::to_u64(&db_info.block_height)?.into(),
                status: db_info.status,
            });
        }
    }

    let prev_block = if let Some(item) = result.last() {
        db_helpers::get_previous_block(pool, item.block_timestamp_nanos.0).await?
    } else {
        return Ok(result);
    };
    let earliest_balance = match super::balance::get_ft_balance_by_contract(
        rpc_client,
        contract_id.clone(),
        account_id.clone(),
        prev_block.height,
    )
    .await
    {
        Ok(x) => x,
        Err(e) => {
            if e.message.contains("does not exist at block_height") {
                0
            } else {
                return Err(e);
            }
        }
    };
    if types::numeric::to_u128(&current_balance)? != earliest_balance {
        return Err(errors::ErrorKind::InternalError(format!(
            "History is not supported for account {}. Contract {} provides inconsistent events",
            account_id, contract_id
        ))
        .into());
    }

    Ok(result)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::modules::tests::*;

    // TODO to support near history tests, we need to recollect the data
    // #[tokio::test]
    // async fn test_near_history() {
    //     let block = get_block();
    //     let pool = init_balances_db().await;
    //     let account = near_primitives::types::AccountId::from_str("vasya.near").unwrap();
    //     let pagination = types::query_params::Pagination {
    //         limit: 3,
    //         after_event_index: None,
    //     };
    //
    //     let history = get_near_history(&pool, &account, &block, &pagination).await;
    //     insta::assert_debug_snapshot!(history);
    // }
    //
    // #[tokio::test]
    // async fn test_near_history_next_page() {
    //     let block = get_block();
    //     let pool = init_balances_db().await;
    //     let account = near_primitives::types::AccountId::from_str("vasya.near").unwrap();
    //     let pagination = types::query_params::Pagination {
    //         limit: 3,
    //         after_event_index: Some(16197887992398226000000000000000003),
    //     };
    //
    //     let history = get_near_history(&pool, &account, &block, &pagination).await;
    //     insta::assert_debug_snapshot!(history);
    // }
    //
    // #[tokio::test]
    // async fn test_near_history_with_failed_receipts() {
    //     let block = db_helpers::Block {
    //         timestamp: 1618591017607373869,
    //         height: 34943083,
    //     };
    //     let pool = init_balances_db().await;
    //     let account = near_primitives::types::AccountId::from_str("zubkowi.near").unwrap();
    //     let pagination = types::query_params::Pagination {
    //         limit: 10,
    //         after_event_index: None,
    //     };
    //
    //     let history = get_near_history(&pool, &account, &block, &pagination).await;
    //     insta::assert_debug_snapshot!(history);
    // }
    //
    // #[tokio::test]
    // async fn test_near_history_account_does_not_exist() {
    //     let block = get_block();
    //     let pool = init_balances_db().await;
    //     let account =
    //         near_primitives::types::AccountId::from_str("two-idiots-and-a-half.near").unwrap();
    //     let pagination = types::query_params::Pagination {
    //         limit: 10,
    //         after_event_index: None,
    //     };
    //
    //     let history = get_near_history(&pool, &account, &block, &pagination)
    //         .await
    //         .unwrap();
    //     assert!(history.is_empty());
    // }

    #[tokio::test]
    async fn test_coin_history() {
        let pool = init_db().await;
        let pool_balances = init_balances_db().await;
        let rpc_client = init_rpc();
        let block = get_block();
        let contract = near_primitives::types::AccountId::from_str("wrap.near").unwrap();
        let account = near_primitives::types::AccountId::from_str("caoducanh98.near").unwrap();
        let pagination = types::query_params::Pagination {
            limit: 10,
            after_event_index: None,
        };

        let history = get_coin_history(
            &pool,
            &pool_balances,
            &rpc_client,
            &contract,
            &account,
            &block,
            &pagination,
        )
        .await;
        insta::assert_debug_snapshot!(history);
    }

    #[tokio::test]
    async fn test_coin_history_next_page() {
        let pool = init_db().await;
        let pool_balances = init_balances_db().await;
        let rpc_client = init_rpc();
        let contract = near_primitives::types::AccountId::from_str("wrap.near").unwrap();
        let account = near_primitives::types::AccountId::from_str("caoducanh98.near").unwrap();
        let pagination = types::query_params::Pagination {
            limit: 5,
            after_event_index: Some(16351410322240000000000000000080003),
        };
        let block = db_helpers::checked_get_block_from_pagination(&pool, &pagination)
            .await
            .unwrap();

        let history = get_coin_history(
            &pool,
            &pool_balances,
            &rpc_client,
            &contract,
            &account,
            &block,
            &pagination,
        )
        .await;
        insta::assert_debug_snapshot!(history);
    }

    #[tokio::test]
    async fn test_coin_history_with_failed_receipts() {
        let pool = init_db().await;
        let pool_balances = init_balances_db().await;
        let rpc_client = init_rpc();
        let block = get_block();
        let contract = near_primitives::types::AccountId::from_str("wrap.near").unwrap();
        let account = near_primitives::types::AccountId::from_str("ulvend.near").unwrap();
        let pagination = types::query_params::Pagination {
            limit: 10,
            after_event_index: None,
        };

        let history = get_coin_history(
            &pool,
            &pool_balances,
            &rpc_client,
            &contract,
            &account,
            &block,
            &pagination,
        )
        .await;
        insta::assert_debug_snapshot!(history);
    }

    #[tokio::test]
    async fn test_coin_history_account_does_not_exist() {
        let pool = init_db().await;
        let pool_balances = init_balances_db().await;
        let rpc_client = init_rpc();
        let block = get_block();
        let contract = near_primitives::types::AccountId::from_str("wrap.near").unwrap();
        let account =
            near_primitives::types::AccountId::from_str("two-idiots-and-a-half.near").unwrap();
        let pagination = types::query_params::Pagination {
            limit: 10,
            after_event_index: None,
        };

        let history = get_coin_history(
            &pool,
            &pool_balances,
            &rpc_client,
            &contract,
            &account,
            &block,
            &pagination,
        )
        .await
        .unwrap();
        assert!(history.is_empty());
    }

    #[tokio::test]
    async fn test_coin_history_contract_does_not_exist() {
        let pool = init_db().await;
        let pool_balances = init_balances_db().await;
        let rpc_client = init_rpc();
        let block = get_block();
        let contract =
            near_primitives::types::AccountId::from_str("two-idiots-and-a-half.near").unwrap();
        let account = near_primitives::types::AccountId::from_str("olga.near").unwrap();
        let pagination = types::query_params::Pagination {
            limit: 10,
            after_event_index: None,
        };

        let balance = get_coin_history(
            &pool,
            &pool_balances,
            &rpc_client,
            &contract,
            &account,
            &block,
            &pagination,
        )
        .await;
        insta::assert_debug_snapshot!(balance);
    }

    // TODO we need to rerun this test and update snapshot when the data will be collected
    // #[tokio::test]
    // async fn test_coin_history_contract_is_inconsistent() {
    //     let pool = init_db().await;
    //     let pool_events = init_events_db().await; let rpc_client = init_rpc();
    //     let block = get_block();
    //     let contract = near_primitives::types::AccountId::from_str("coin.asac.near").unwrap();
    //     let account = near_primitives::types::AccountId::from_str("olga.near").unwrap();
    //     let pagination = types::query_params::Pagination {
    //         limit: 10,
    //         after_event_index: None,
    //     };
    //
    //     let balance =
    //         get_coin_history(&pool, &pool_events, &rpc_client, &contract, &account, &block, &pagination).await;
    //     insta::assert_debug_snapshot!(balance);
    // }
}
