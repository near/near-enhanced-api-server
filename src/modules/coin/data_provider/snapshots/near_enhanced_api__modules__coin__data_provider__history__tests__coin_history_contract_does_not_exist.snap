---
source: src/modules/coin/data_provider/history.rs
expression: balance
---
Err(
    Error {
        code: 400,
        message: "Invalid Input: account_id two-idiots-and-a-half.near does not exist at block_height 68000000",
        retriable: false,
    },
)
